To run all the unit test cases in test_cast.py, the TestCast12 will be failed, but to run the TestCast12 case seperately, it will be passed. there should be a bug.
root@3FG16:/workspace/dpdp_habana/PaddleCustomDevice/backends/intel_hpu/tests/unittests# python test_cast.py
/usr/local/lib/python3.10/dist-packages/paddle/utils/cpp_extension/extension_utils.py:686: UserWarning: No ccache found. Please be aware that recompiling all source files may be required. You can download and install ccache from: https://github.com/ccache/ccache/blob/master/doc/INSTALL.md
  warnings.warn(warning_message)
I1224 16:24:43.060763 15055 init.cc:235] ENV [CUSTOM_DEVICE_ROOT]=/usr/local/lib/python3.10/dist-packages/paddle_custom_device
I1224 16:24:43.060786 15055 init.cc:144] Try loading custom device libs from: [/usr/local/lib/python3.10/dist-packages/paddle_custom_device]
I1224 16:24:43.152319 15055 custom_device.cc:1099] Succeed in loading custom runtime in lib: /usr/local/lib/python3.10/dist-packages/paddle_custom_device/libpaddle-intel-hpu.so
I1224 16:24:43.153618 15055 custom_kernel.cc:63] Succeed in loading 107 custom kernel(s) from loaded lib(s), will be used like native ones.
I1224 16:24:43.153707 15055 init.cc:156] Finished in LoadCustomDevice with libs_path: [/usr/local/lib/python3.10/dist-packages/paddle_custom_device]
I1224 16:24:43.153721 15055 init.cc:241] CustomDevice: intel_hpu, visible devices count: 8
I1224 16:25:33.004395 15055 program_interpreter.cc:243] New Executor is Running.
..F..................I1224 16:39:08.336874 15055 interpreter_util.cc:651] Standalone Executor is Used.
.
======================================================================
FAIL: test_check_output (__main__.TestCast12)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/workspace/dpdp_habana/PaddleCustomDevice/backends/intel_hpu/tests/unittests/test_cast.py", line 65, in test_check_output
    self.check_output_with_place(self.place)
  File "/workspace/dpdp_habana/PaddleCustomDevice/python/tests/op_test.py", line 2727, in check_output_with_place
    static_checker.check()
  File "/workspace/dpdp_habana/PaddleCustomDevice/python/tests/op_test.py", line 2329, in check
    self.compare_outputs_with_expects()
  File "/workspace/dpdp_habana/PaddleCustomDevice/python/tests/op_test.py", line 2319, in compare_outputs_with_expects
    self.compare_single_output_with_expect(out_name, expect)
  File "/workspace/dpdp_habana/PaddleCustomDevice/python/tests/op_test.py", line 2299, in compare_single_output_with_expect
    self._compare_numpy(name, actual_np, expect_np)
  File "/workspace/dpdp_habana/PaddleCustomDevice/python/tests/op_test.py", line 2266, in _compare_numpy
    np.testing.assert_allclose(
  File "/usr/local/lib/python3.10/dist-packages/numpy/testing/_private/utils.py", line 1678, in assert_allclose
    assert_array_compare(compare, actual, desired, err_msg=str(err_msg),
  File "/usr/local/lib/python3.10/dist-packages/numpy/testing/_private/utils.py", line 884, in assert_array_compare
    raise AssertionError(msg)
AssertionError:
Not equal to tolerance rtol=0.01, atol=0.03
Operator (cast) Output (Out) has diff at Place(intel_hpu:0) in static checker
Mismatched elements: 32768000 / 32768000 (100%)
Max absolute difference among violations: 2.
Max relative difference among violations: inf
ACTUAL: array([[1.695312, 1.28125 , 1.226562, ..., 1.789062, 1.84375 , 1.054688],
       [1.539062, 1.695312, 1.625   , ..., 1.546875, 1.679688, 1.132812],
       [1.265625, 1.484375, 1.765625, ..., 1.351562, 1.0625  , 1.835938],...
DESIRED: array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],...
 
----------------------------------------------------------------------
Ran 22 tests in **864.985s**
 
FAILED (failures=1)

root@3FG16:/workspace/dpdp_habana/PaddleCustomDevice/backends/intel_hpu/tests/unittests# python test_cast.py TestCast12 -v
/usr/local/lib/python3.10/dist-packages/paddle/utils/cpp_extension/extension_utils.py:686: UserWarning: No ccache found. Please be aware that recompiling all source files may be required. You can download and install ccache from: https://github.com/ccache/ccache/blob/master/doc/INSTALL.md
  warnings.warn(warning_message)
I1224 16:41:13.106348 15332 init.cc:235] ENV [CUSTOM_DEVICE_ROOT]=/usr/local/lib/python3.10/dist-packages/paddle_custom_device
I1224 16:41:13.106367 15332 init.cc:144] Try loading custom device libs from: [/usr/local/lib/python3.10/dist-packages/paddle_custom_device]
I1224 16:41:13.195590 15332 custom_device.cc:1099] Succeed in loading custom runtime in lib: /usr/local/lib/python3.10/dist-packages/paddle_custom_device/libpaddle-intel-hpu.so
I1224 16:41:13.196866 15332 custom_kernel.cc:63] Succeed in loading 107 custom kernel(s) from loaded lib(s), will be used like native ones.
I1224 16:41:13.196951 15332 init.cc:156] Finished in LoadCustomDevice with libs_path: [/usr/local/lib/python3.10/dist-packages/paddle_custom_device]
I1224 16:41:13.196966 15332 init.cc:241] CustomDevice: intel_hpu, visible devices count: 8
test_check_output (__main__.TestCast12) ... I1224 16:42:03.336819 15332 program_interpreter.cc:243] New Executor is Running.
ok
 
----------------------------------------------------------------------
Ran 1 test in 124.973s

--- a/backends/intel_hpu/tests/unittests/test_cast.py
+++ b/backends/intel_hpu/tests/unittests/test_cast.py
@@ -84,5 +84,211 @@ class TestCastBF16_2(TestCastBF16_1):
         self.out_dtype = int(core.VarDesc.VarType.FP32)
 
 
+class TestCastBF16_3(TestCastBF16):
+    def init_shape(self):
+        self.shape = [2, 1, 1, 4096]
+
+    def init_dtype(self):
+        self.input_dtype = "bool"
+        self.output_dtype = "uint16"
+        self.in_dtype = int(core.VarDesc.VarType.BOOL)
+        self.out_dtype = int(core.VarDesc.VarType.BF16)
+
+
+class TestCastBF16_4(TestCastBF16):
+    def init_shape(self):
+        self.shape = [1]
+
+
+class TestCastBF16_5(TestCastBF16):
+    def init_shape(self):
+        self.shape = [1024, 8192]
+
+
+@skip_check_grad_ci(reason="[skip NPU cast grad check] not implemented yet.")
+class TestCastBF16_5(TestCastBF16):
+    def init_dtype(self):
+        self.input_dtype = "float16"
+        self.output_dtype = "float32"
+        self.in_dtype = int(core.VarDesc.VarType.FP16)
+        self.out_dtype = int(core.VarDesc.VarType.FP32)
+
+    def test_check_output(self):
+        self.check_output_with_place(self.place, atol=1e-3)
+
+
+@skip_check_grad_ci(reason="[skip NPU cast grad check] not implemented yet.")
+class TestCastBF16_6(TestCastBF16):
+    def init_dtype(self):
+        self.input_dtype = "int32"
+        self.output_dtype = "int32"
+        self.in_dtype = int(core.VarDesc.VarType.INT32)
+        self.out_dtype = int(core.VarDesc.VarType.INT32)
+
+    def test_check_output(self):
+        self.check_output_with_place(self.place, atol=1e-3)
+
+
+class TestCastBF16_7(TestCastBF16):
+    def init_shape(self):
+        self.shape = [2, 4096, 1]
+
+    def init_dtype(self):
+        self.input_dtype = "float"
+        self.output_dtype = "bool"
+        self.in_dtype = int(core.VarDesc.VarType.FP32)
+        self.out_dtype = int(core.VarDesc.VarType.BOOL)
+
+
+class TestCastBF16_8(TestCastBF16):
+    def init_shape(self):
+        self.shape = [4096, 4096]
+
+    def init_dtype(self):
+        self.input_dtype = "float"
+        self.output_dtype = "int32"
+        self.in_dtype = int(core.VarDesc.VarType.FP32)
+        self.out_dtype = int(core.VarDesc.VarType.INT32)
+
+
+class TestCastBF16_9(TestCastBF16):
+    def init_shape(self):
+        self.shape = [8192]
+
+    def init_dtype(self):
+        self.input_dtype = "int64"
+        self.output_dtype = "bool"
+        self.in_dtype = int(core.VarDesc.VarType.INT64)
+        self.out_dtype = int(core.VarDesc.VarType.BOOL)
+
+
+class TestCastBF16_10(TestCastBF16):
+    def init_shape(self):
+        self.shape = [2, 1, 1, 4096]
+
+    def init_dtype(self):
+        self.input_dtype = "bool"
+        self.output_dtype = "uint16"
+        self.in_dtype = int(core.VarDesc.VarType.BOOL)
+        self.out_dtype = int(core.VarDesc.VarType.BF16)
+
+
+class TestCast10(TestCastBF16):
+    def init_shape(self):
+        self.shape = [2, 4096, 4000]
+
+    def init_dtype(self):
+        self.input_dtype = "float"
+        self.output_dtype = "uint16"
+        self.in_dtype = int(core.VarDesc.VarType.FP32)
+        self.out_dtype = int(core.VarDesc.VarType.BF16)
+
+
+class TestCast11(TestCast10):
+    def init_shape(self):
+        self.shape = [3584, 8192]
+
+
+class TestCast12(TestCast10):
+    def init_shape(self):
+        self.shape = [4000, 8192]
+
+
+class TestCast13(TestCast10):
+    def init_shape(self):
+        self.shape = [8192, 1280]
+
+
+class TestCast14(TestCast10):
+    def init_shape(self):
+        self.shape = [8192, 7168]
+
+
+class TestCast15(TestCastBF16):
+    def init_dtype(self):
+        self.input_dtype = "int64"
+        self.output_dtype = "float32"
+        self.in_dtype = int(core.VarDesc.VarType.INT64)
+        self.out_dtype = int(core.VarDesc.VarType.FP32)
+
+    def init_shape(self):
+        self.shape = [8192, 1]
+
+
+class TestCast16(TestCastBF16):
+    def init_dtype(self):
+        self.input_dtype = "bool"
+        self.output_dtype = "float32"
+        self.in_dtype = int(core.VarDesc.VarType.BOOL)
+        self.out_dtype = int(core.VarDesc.VarType.FP32)
+
+    def init_shape(self):
+        self.shape = [8192, 4000]
+
+
+class TestCastOpFp32ToFp64(OpTest):
+    def setUp(self):
+        self.set_npu()
+        self.op_type = "cast"
+        self.place = paddle.CustomPlace("intel_hpu", 0)
+
+        ipt = np.random.random(size=[10, 10])
+        self.inputs = {"X": ipt.astype("float32")}
+        self.outputs = {"Out": ipt.astype("float64")}
+        self.attrs = {
+            "in_dtype": int(core.VarDesc.VarType.FP32),
+            "out_dtype": int(core.VarDesc.VarType.FP64),
+        }
+
+    def set_npu(self):
+        self.__class__.use_custom_device = True
+
+    def test_check_output(self):
+        self.check_output()
+
+    def test_grad(self):
+        self.check_grad(["X"], ["Out"])
+
+
+class TestCastOpFp16ToFp32(OpTest):
+    def setUp(self):
+        self.set_npu()
+        ipt = np.random.random(size=[10, 10])
+        self.inputs = {"X": ipt.astype("float16")}
+        self.outputs = {"Out": ipt.astype("float32")}
+        self.attrs = {
+            "in_dtype": int(core.VarDesc.VarType.FP16),
+            "out_dtype": int(core.VarDesc.VarType.FP32),
+        }
+        self.op_type = "cast"
+        self.__class__.no_need_check_grad = True
+
+    def set_npu(self):
+        self.__class__.use_custom_device = True
+
+    def test_check_output(self):
+        self.check_output(atol=1e-3)
+
+
+class TestCastOpBf16ToFp32(OpTest):
+    def setUp(self):
+        self.set_npu()
+        ipt = np.array(np.random.randint(10, size=[10, 10])).astype("uint16")
+        self.inputs = {"X": ipt}
+        self.outputs = {"Out": convert_uint16_to_float(ipt)}
+        self.attrs = {
+            "in_dtype": int(core.VarDesc.VarType.BF16),
+            "out_dtype": int(core.VarDesc.VarType.FP32),
+        }
+        self.op_type = "cast"
+        self.__class__.no_need_check_grad = True
+
+    def set_npu(self):
+        self.__class__.use_custom_device = True
+
+    def test_check_output(self):
+        self.check_output()
+
+
 if __name__ == "__main__":
     unittest.main()

